Feature/corr table changes

zsloan opened the pull request:
#### Description
This PR is intended to implement Scroller with the correlation page table and convert that correlation page table to using JSON input instead of drawing all the HTML and then converting that to DataTables.

It will also change the way the show/hide column buttons work to avoid the glossary link question marks being included in them.

#### How should this be tested?
Do correlations from any trait page

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
N/A

#### Screenshots (if appropriate)

#### Questions

zsloan committed:
Imported Scroller and used it with correlation page table

zsloan committed:
Removed html from generate_corr_json since that will be done by the column definitions after passing the data to DataTables as JSON

zsloan committed:
Added some additional necessary fields to generate_corr_json in show_corr_results.py

zsloan committed:
Converted correlation table for ProbeSet target datasets to use the JSON input and removed the unused table html

zsloan committed:
Added CSS file for all pages that include a table of traits

zsloan committed:
* wqflask/wqflask/correlation/show_corr_results.py - Changed some of the parameters sent to the correlation table
* wqflask/wqflask/templates_correlation_page.html - Fixed some column defs and removed unused table configuration code

zsloan committed:
Switched checking if a trait is confidential to checking if it has a pubmed ID for the purposes of showing pre-publication abbreviation/description

zsloan committed:
- Fixed an issue causing the phenotype abbreviation to not be displayed properly in correlation results
- Set default description to N/A when there's no trait description for phenotype traits

zsloan committed:
Changed some column widths and removed the width for description so it can vary with page width

zsloan committed:
Merge branch 'testing' of github.com:genenetwork/genenetwork2 into feature/corr_table_changes

zsloan committed:
Replaced superscript i's for glossary links in search result table
headers with fontawesome icons

zsloan committed:
Added cdn imports for fontawesome to correlation page and changed
superscript i's to fontawesome icons for the glossary links in column
headers

BonfaceKilz commented on wqflask/wqflask/correlation/show_corr_results.py:
```diff
@@ -484,30 +484,31 @@ def generate_corr_json(corr_results, this_trait, dataset, target_dataset, for_ap
             results_dict['description'] = trait.description_display
             results_dict['location'] = trait.location_repr
-            results_dict['mean'] = float(trait.mean)
+            if trait.mean and trait.mean != "":
```
**suggestion:nitpick:** You could cast that into a bool:

``` python
#+begin_src python :results output
print(bool(""))

print(bool(None))

print(bool("some value"))

print(bool(2))
#+end_src

#+RESULTS:
: False
: False
: True
: True
```

So this should work:
```suggestion
            if bool(trait.mean) 
```

BonfaceKilz commented on wqflask/wqflask/correlation/show_corr_results.py:
```diff
@@ -484,30 +484,31 @@ def generate_corr_json(corr_results, this_trait, dataset, target_dataset, for_ap
                 results_dict['lrs_score'] = "N/A"
             results_dict['lrs_location'] = trait.LRS_location_repr
-            if trait.additive != "":
+            if trait.additive and trait.additive != "":
```
See above

zsloan committed:
Changed some if/else statements to just initializing the variables + the
if statement and fixed "lrs_location" to "location" for target genotypes

BonfaceKilz commented on wqflask/wqflask/correlation/show_corr_results.py:
```diff
@@ -484,30 +484,31 @@ def generate_corr_json(corr_results, this_trait, dataset, target_dataset, for_ap
-                results_dict['sample_r'] = "%0.3f" % float(trait.sample_r)
-            else:
-                results_dict['sample_r'] = "<a target='_blank' href='corr_scatter_plot?dataset_1=" + str(dataset.name) + "&dataset_2=" + str(trait.dataset.name) + "&trait_1=" + str(this_trait.name) + "&trait_2=" + str(trait.name) + "'>" + "%0.3f" % float(trait.sample_r) + "</a>"
+            results_dict['sample_r'] = "%0.3f" % float(trait.sample_r)
```
**suggestion:nitpick:** Since we now use Python3, it's a nice idea to switch to [f-strings](http://zetcode.com/python/fstring):

```python
#+begin_src python :results output
float_string = "0.35535115"
print("%0.3f" % float(float_string))

print(f"{float(float_string):.3f}")
#+end_src

#+RESULTS:
: 0.355
: 0.355
```

So this would be:
```suggestion
            results_dict['sample_r'] = f"{float(trait.sample_r):.3f}"
```
Idea is to not do this *at* once. Instead, when we encounter a %-string(or when you need formatted strings), we simply default to f-strings.
cc @Alexanderlacuna

BonfaceKilz commented on wqflask/wqflask/correlation/show_corr_results.py:
```diff
@@ -516,41 +517,49 @@ def generate_corr_json(corr_results, this_trait, dataset, target_dataset, for_ap
             results_dict['lrs_score'] = trait.LRS_score_repr
             results_dict['lrs_location'] = trait.LRS_location_repr
-            if trait.additive != "":
+            if trait.additive and trait.additive != "":
```
See previous comments...

BonfaceKilz commented on wqflask/wqflask/static/new/css/trait_list.css:
```diff
@@ -0,0 +1,34 @@
+    margin-top: 10px;
+}
+
+/* div.dts_label {
+    position: absolute;
+    right: 10px;
+    background: rgba(0, 0, 0, 0.8);
+    color: white;
+    box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.5);
+    text-align: right;
+    border-radius: 3px;
+    padding: 0.4em;
+    z-index: 2;
+    display: none;
+} */
```
**suggestion:** Remove commented out blocks at the very end of this PR

Commented out blocks usually wind up getting stale and sometimes make it confusing to read the code...

BonfaceKilz commented on wqflask/wqflask/static/new/css/trait_list.css:
```diff
@@ -0,0 +1,34 @@
+} */
+
+table.dataTable tbody td {
+    /* word-wrap: break-word;
+    word-break: break-all; */
```
See above.

zsloan commented:
Thanks for the suggestions! I didn't think of using boolean casting that
way and forgot about the f-string syntax.

On Fri, Nov 13, 2020 at 2:30 PM BonfaceKilz <notifications@github.com>
wrote:

> *@BonfaceKilz* requested changes on this pull request.
> ------------------------------
>
> In wqflask/wqflask/correlation/show_corr_results.py
> <https://github.com/genenetwork/genenetwork2/pull/482#discussion_r523196872>
> :
>
> >              results_dict['description'] = trait.description_display
>
>              results_dict['location'] = trait.location_repr
>
> -            results_dict['mean'] = float(trait.mean)
>
> +            if trait.mean and trait.mean != "":
>
>
> *suggestion:nitpick:* You could cast that into a bool:
>
> #+begin_src python :results output
> print(bool(""))
>
>
> print(bool(None))
>
>
> print(bool("some value"))
>
>
> print(bool(2))
> #+end_src
>
>
> #+RESULTS:
>
> : False
>
> : False
>
> : True
>
> : True
>
> So this should work:
> ⬇️ Suggested change
>
> -            if trait.mean and trait.mean != "":
>
> +            if bool(trait.mean)
>
>
> ------------------------------
>
> In wqflask/wqflask/correlation/show_corr_results.py
> <https://github.com/genenetwork/genenetwork2/pull/482#discussion_r523197105>
> :
>
> >              if trait.LRS_score_repr != "N/A":
>
>                  results_dict['lrs_score'] = "%.1f" % float(trait.LRS_score_repr)
>
>              else:
>
>                  results_dict['lrs_score'] = "N/A"
>
>              results_dict['lrs_location'] = trait.LRS_location_repr
>
> -            if trait.additive != "":
>
> +            if trait.additive and trait.additive != "":
>
>
> See above
> ------------------------------
>
> In wqflask/wqflask/correlation/show_corr_results.py
> <https://github.com/genenetwork/genenetwork2/pull/482#discussion_r523200989>
> :
>
> >                  results_dict['additive'] = "%0.3f" % float(trait.additive)
>
>              else:
>
>                  results_dict['additive'] = "N/A"
>
> -            if for_api:
>
> -                results_dict['sample_r'] = "%0.3f" % float(trait.sample_r)
>
> -            else:
>
> -                results_dict['sample_r'] = "<a target='_blank' href='corr_scatter_plot?dataset_1=" + str(dataset.name) + "&dataset_2=" + str(trait.dataset.name) + "&trait_1=" + str(this_trait.name) + "&trait_2=" + str(trait.name) + "'>" + "%0.3f" % float(trait.sample_r) + "</a>"
>
> +            results_dict['sample_r'] = "%0.3f" % float(trait.sample_r)
>
>
> *suggestion:nitpick:* Since we now use Python3, it's a nice idea to
> switch to f-strings <http://zetcode.com/python/fstring>:
>
> #+begin_src python :results output
> float_string = "0.35535115"
> print("%0.3f" % float(float_string))
>
>
> print(f"{float(float_string):.3f}")
> #+end_src
>
>
> #+RESULTS:
>
> : 0.355
>
> : 0.355
>
> So this would be:
> ⬇️ Suggested change
>
> -            results_dict['sample_r'] = "%0.3f" % float(trait.sample_r)
>
> +            results_dict['sample_r'] = f"{float(trait.sample_r):.3f}"
>
>
> Idea is to not do this *at* once. Instead, when we encounter a
> %-string(or when you need formatted strings), we simply default to
> f-strings.
> cc @Alexanderlacuna <https://github.com/Alexanderlacuna>
> ------------------------------
>
> In wqflask/wqflask/correlation/show_corr_results.py
> <https://github.com/genenetwork/genenetwork2/pull/482#discussion_r523201223>
> :
>
> >              results_dict['lrs_score'] = trait.LRS_score_repr
>
>              results_dict['lrs_location'] = trait.LRS_location_repr
>
> -            if trait.additive != "":
>
> +            if trait.additive and trait.additive != "":
>
>
> See previous comments...
> ------------------------------
>
> In wqflask/wqflask/static/new/css/trait_list.css
> <https://github.com/genenetwork/genenetwork2/pull/482#discussion_r523201739>
> :
>
> > +/* div.dts_label {
>
> +    position: absolute;
>
> +    right: 10px;
>
> +    background: rgba(0, 0, 0, 0.8);
>
> +    color: white;
>
> +    box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.5);
>
> +    text-align: right;
>
> +    border-radius: 3px;
>
> +    padding: 0.4em;
>
> +    z-index: 2;
>
> +    display: none;
>
> +} */
>
>
> *suggestion:* Remove commented out blocks at the very end of this PR
>
> Commented out blocks usually wind up getting stale and sometimes make it
> confusing to read the code...
> ------------------------------
>
> In wqflask/wqflask/static/new/css/trait_list.css
> <https://github.com/genenetwork/genenetwork2/pull/482#discussion_r523201836>
> :
>
> > +    /* word-wrap: break-word;
>
> +    word-break: break-all; */
>
>
> See above.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/482#pullrequestreview-530371456>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANQJGFPEL6IQ6HOWLDBKNTSPWJMRANCNFSM4TL7KYTQ>
> .
>

Alexanderlacuna commented:
@BonfaceKilz Thanks for the suggestions.

zsloan committed:
Added some css to trait_list.css restyling the "show/hide column"
buttons when they're active. Not really pleased with the result, but
styling a button to look depressed is a pain

zsloan committed:
- Added Show/Hide Columns options and included the trait_list.css file
which including styling for said options
- Changed some column widths to avoid text wrapping

zsloan committed:
Switched to using boolean casting to check if various trait attributes
exist

zsloan committed:
Switched to using f-strings instead of the older method of string
formatting

zsloan committed:
Fixed the "Reset Columns" button to work with the new way of showing/hiding columns (that no longer uses DataTables buttons plug-in)

zsloan commented:
I think this should be ready now? I'll do separate PRs for converting the other tables to using Scroller in the same way.
